### PR TITLE
Upper boundary conditions specification via namelist

### LINKED ISSUE #216: Need to expose species upper boundary conditions as namelist options
In WACCM (and WACCMX) how upper boundary conditions need to be explicitly set in the namelist rather than hardwired and buried in the code.

### LINKED ISSUE #533: Option to specify upper boundary conditions from a file
It would be advantageous to be able to specify an upper boundary condition in CAM, CAM-CHEM and WACCM from a file. CAM-CHEM currently has a zero flux condition and so misses upper atmospheric sources of NOy important for ozone in the stratosphere. Similarly, CAM and CAM-CHEM in "low top" configurations likely miss sources of water vapor that come from methane oxidation in the uppermost stratosphere and mesosphere. In the "high top" CAM version (lid ~80km) specification of mesopause temperature and water vapor, CO2, etc. is needed rather than a hard coded value that does not change with climate change/emissions. In WACCM some values at the UBC are taken from a TIEGCM file or specified from empirical models (MSIS, NOEM) but could easily be set to use an WACCM-X UBC.

I am not sure what the best way to implement this is, but since WACCM has the ability to specify it's UBC from a TIEGCM file then perhaps that code can be generalized.

I see the main source of input UBCs would likely be CMIP6 timeseries files from simulations done by WACCM, mostly likely zonal-mean monthly data.

### PR DESCRIPTION
**Pull request by fvitt**
Implements upper boundary conditions namelist options for low and high-top models.  The defaults setting only specify fixed upper boundary conditions for WACCM.

closes #216 
closes #533

### CODE REVIEW

**Comment by fvitt** [comment]
Should remove deprecated NOy UBC module (dead code):
 src/chemistry/mozart/noy_ubc.F90

**Comment by cacraigucar** [comment]
> Should remove deprecated NOy UBC module (dead code): src/chemistry/mozart/noy_ubc.F90

@fvitt - Feel free to make this change and push it back to your branch.  I started reviewing yesterday, but this change won't impact my review.

**Review by cacraigucar — CHANGES_REQUESTED** [review]

File: doc/ChangeLog [hunk 1]
```diff
@@ -1,4 +1,159 @@
...
+ - protect cnst_fixed_ubc and cnst_fixed_ubflx arrays
+
+M	src/physics/cam/diffusion_solver.F90
+ - explicitly set mmr in top layer for cases where molecular diffusion is not active
+
+M	src/physics/cam/physpkg.F90
+ - determine if water vapor constituent has fixed UBC
+
+M	src/physics/cam/vertical_diffusion.F90
+ - get upper boundar fluxes separately for WACCMX
```
[the remark this answers is not in the document]
Comment (fvitt) [reply doc/ChangeLog:116]: fixed

File: src/chemistry/mozart/upper_bc_file.F90 [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+          end if
+       end if
+       close(unitn)
+    end if
+
+    call mpi_bcast(ubc_file_path, len(ubc_file_path), mpi_character, masterprocid, mpicom, ierr)
+    call mpi_bcast(ubc_file_input_type, len(ubc_file_input_type), mpi_character, masterprocid, mpicom, ierr)
+    call mpi_bcast(ubc_file_fixed_ymd, 1, mpi_integer, masterprocid, mpicom, ierr)
+    call mpi_bcast(ubc_file_fixed_tod, 1, mpi_integer, masterprocid, mpicom, ierr)
+    call mpi_bcast(ubc_file_cycle_yr,  1, mpi_integer, masterprocid, mpicom, ierr)
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/mozart/upper_bc_file.F90]: added checks

File: src/chemistry/mozart/upper_bc.F90 [hunk 3]
```diff
@@ -1,150 +1,197 @@
...
+    call mpi_bcast(tgcm_ubc_cycle_yr, 1, mpi_integer, masterprocid, mpicom, ierr)
+    call mpi_bcast(tgcm_ubc_fixed_ymd, 1, mpi_integer, masterprocid, mpicom, ierr)
+    call mpi_bcast(tgcm_ubc_fixed_tod, 1, mpi_integer, masterprocid, mpicom, ierr)
+    call mpi_bcast(snoe_ubc_file, len(snoe_ubc_file), mpi_character, masterprocid, mpicom, ierr)
+    call mpi_bcast(t_pert_ubc, 1, mpi_real8, masterprocid, mpicom, ierr)
+    call mpi_bcast(no_xfac_ubc,1, mpi_real8, masterprocid, mpicom, ierr)
+    call mpi_bcast(ubc_specifier, pcnst*len(ubc_specifier(1)), mpi_character,masterprocid, mpicom, ierr)
+    call mpi_bcast(ubc_flds,      pcnst*len(ubc_flds(1)),      mpi_character,masterprocid, mpicom, ierr)
+    call mpi_bcast(ubc_file_spfr, pcnst*len(ubc_file_spfr(1)), mpi_character,masterprocid, mpicom, ierr)
+    call mpi_bcast(ubc_source,    pcnst*len(ubc_source(1)),    mpi_character,masterprocid, mpicom, ierr)
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/mozart/upper_bc.F90]: added checks

File: src/chemistry/mozart/upper_bc.F90 [hunk 4]
```diff
@@ -157,39 +204,159 @@ subroutine ubc_init()
...
+       hist_names(m) = trim(ubc_flds(m))//'_UBC'
+       if (ubc_flds(m)=='T') then
+          ubc_fixed_temp=.true.
+          spc_ndx(m) = -1
+          call addfld(hist_names(m), horiz_only, 'I', 'K', trim(ubc_flds(m))//' at upper boundary' )
+       else
+          call cnst_get_ind(ubc_flds(m), spc_ndx(m), abort=.true.)
+          call addfld(hist_names(m), horiz_only, 'I', 'kg/kg', trim(ubc_flds(m))//' at upper boundary' )
+       end if
+       !call add_default(hist_names(m), 2, ' ' )
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/mozart/upper_bc.F90]: removed

**Review by cacraigucar — APPROVED** [review]

File: src/chemistry/mozart/upper_bc.F90 [hunk 5]
```diff
@@ -356,22 +491,82 @@ subroutine ubc_get_vals (lchnk, ncol, pint, zi, t, q, omega, phis, &
...
+       !---------------------------------------------------------------------
+       zkt = (rEarth + ( 0.5_r8 * ( zi(iCol,1) + zi(iCol,2) ) + rga * phis(iCol) ) ) * &
+            cnst_mw(h_ndx) / avogad * grav / ( kboltz * t(iCol,1) )
+
+       ubc_flux(iCol,h_ndx) = hfluxlimitfac * SQRT(kboltz/(2.0_r8 * pi * cnst_mw(h_ndx) / avogad)) * &
+            qh_top(iCol) * nmbartop * &
+            SQRT(t(iCol,1)) * (1._r8 + zkt) * EXP(-zkt)
+
+       ubc_flux(iCol,h_ndx) = ubc_flux(iCol,h_ndx) * &
+            (2.03E-13_r8 * qh_top(iCol) * nmbartop / (cnst_mw(h_ndx) / avogad) * t(iCol,1))
```
Comment (nusbaume) [inline src/chemistry/mozart/upper_bc.F90:566]: The number `2.03E-13` looks really specific.  Do you know where it came from?

**Review by nusbaume — CHANGES_REQUESTED** [review]
Looks good, but I did have some (hopefully minor) change requests.

File: src/chemistry/mozart/upper_bc.F90 (hunk 5 above)
Comment (fvitt) [reply src/chemistry/mozart/upper_bc.F90:566]: @hanli-liu Can you comment on the `2.03E-13` factor used in the calculation of the fluxes of H at the upper boundary of WACCMX?

Comment (hanli-liu) [reply src/chemistry/mozart/upper_bc.F90:566]: This is the hydrogen escape flux due to charge exchange adopted from TIME-GCM. According to TIME-GCM notes, this value comes from Liu & Donahue (1974) and Hunten & Donahue (1976).

Comment (nusbaume) [reply src/chemistry/mozart/upper_bc.F90:566]: Thanks @hanli-liu!  Do the TIME-GCM notes have the titles for the papers as well?  It looks like there were multiple "Liu & Donahue 1974" papers related to Hydrogen in the upper atmosphere, so I just want to make sure we cite the correct one.

Comment (nusbaume) [reply src/chemistry/mozart/upper_bc.F90:566]: I am also happy with instead just using the TIME-GCM notes directly if they have been published somewhere (whichever option is easier).

Comment (hanli-liu) [reply src/chemistry/mozart/upper_bc.F90:566]: You may use the following reference for TIME-GCM:
Roble, R. G., Energetics of the mesosphere and thermosphere, in The Upper Mesosphere and Lower Thermosphere: A Review of Experiment and Theory, Geophys. Monogr. Ser., vol. 87, edited by R. M. Johnson and T. L. Killeen, pp. 1 –22, AGU, Washington, D. C., 1995.
This paper does not give a detailed description of the flux used in the model, but it did refer to the work by Liu&Donahue (1974) and Hunten&Donahue (1976). The Liu&Donahue reference is:
Liu, S.C., and T. M. Donahue, Mesospheric hydrogen related to exospheric escape mechanisms, J. Atmos. Sci.,
31, 1466-1470, 1974. (Equation 4 there).

Comment (fvitt) [reply src/chemistry/mozart/upper_bc.F90:566]: @nusbaume and @hanli-liu 
Are you okay with making this factor a parameter documented as:
```
    real(r8), parameter :: h_escape_flx_factor = 2.03e-13_r8 ! for hydrogen escape flux due to charge exchange
    ! adopted from TIME-GCM (T. L. Killeen, pp. 1 –22, AGU, Washington, D. C., 1995) following
    ! Liu, S.C., and T. M. Donahue, Mesospheric hydrogen related to exospheric escape mechanisms, J. Atmos. Sci.,
    ! 31, 1466-1470, 1974. (Equation 4 there). DOI: 10.1175/1520-0469(1974)031<1466:Mhrtee>2.0.Co;2
    ! https://journals.ametsoc.org/view/journals/atsc/31/5/1520-0469_1974_031_1466_mhrtee_2_0_co_2.xml
```

Comment (hanli-liu) [reply src/chemistry/mozart/upper_bc.F90:566]: The reference to TIME-GCM:
R. G. Roble, pp. 1-21, AGU Geophys. Monogr. Ser 87, 1995.

Comment (nusbaume) [reply src/chemistry/mozart/upper_bc.F90:566]: @fvitt Making the factor a parameter with the provided documentation sounds good to me.  Thanks!
@hanli-liu Thanks again for looking all of this up!

File: src/chemistry/mozart/upper_bc.F90 [hunk 6]
```diff
@@ -1,150 +1,212 @@
...
+
+          if (trim(ubc_source(n))=='ubc_file') then
+             ubc_file_spfr(m) = trim(ubc_specifier(n)(:ndx_ar-1))
+             m=m+1
+          endif
+          if (index(ubc_source(n),'mmr')>0) then
+             n_fixed_mmr=n_fixed_mmr+1
+          end if
+          if (index(ubc_source(n),'vmr')>0) then
+             n_fixed_vmr=n_fixed_vmr+1
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply src/chemistry/mozart/upper_bc.F90:152]: Looks like this question was missed from the first review round.

File: bld/build-namelist [hunk 7]
```diff
@@ -2601,6 +2601,17 @@ if ($maxzen>0.0) {
     add_default($nl, 'photo_max_zen', 'val'=>"${maxzen}D0");
 }
 
+# upper boundary specifier
+my $ubc_ver;
+if (($chem =~ /waccm_ma/ or $chem =~ /waccm_tsmlt/ ) and !$waccmx) {
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply bld/build-namelist:2606]: Looks like this question was missed from the first review round.

**Review by nusbaume — APPROVED** [review]
Looks good now, although there were a couple of questions missed from the first review round that I tried to highlight (but I don't think they have to hold up this PR).

File: doc/ChangeLog [hunk 8]
```diff
@@ -1,4 +1,172 @@
...
+
+Purpose of changes (include the issue number and title text for each relevant GitHub issue):
+
+ Address github issues:
+ . Need to expose species upper boundary conditions as namelist options #216
+ . Option to specify upper boundary conditions from a file #533
+
+ Remove obsolete noy_ubc module
+
+Describe any changes made to build system:
```
Comment (gold2718) [inline doc/ChangeLog:17]: Please indicate no changes:
```suggestion
Describe any changes made to build system: NA
```

File: doc/ChangeLog [hunk 9]
```diff
@@ -1,4 +1,172 @@
...
+ New namelist options for specifying UBCs:
+
+ ubc_specifier:
+   List of fields that have fixed upper boundary conditions (temperature and mass mixing ratios)
+   and corresponding data sources. Each entry in this list is of the form:
+     FLD->source
+   where source may be of the following:
+     MSIS
+     TGCM
+     SNOE
```
Comment (gold2718) [inline doc/ChangeLog:30]: I do not think you need to repeat all the details of the namelist changes here, however, if you do keep it, please update this section if (when?) you update the description in namelist_definition.xml

**Review by gold2718 — CHANGES_REQUESTED** [review]
Just a few requests and questions.

File: bld/build-namelist (hunk 7 above)
[the remark this answers is not in the document]
Comment (fvitt) [reply bld/build-namelist:2606]: Not for WACCM-X and maybe not for CAM (not sure). I added a note to the description.

File: src/chemistry/mozart/upper_bc.F90 (hunk 6 above)
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/mozart/upper_bc.F90]: If both `mmr` and `vmr` are both present (or some other unrecognized string is used) then at initialization the model will abort with an error message something like:
 `ERROR: ubc_init: incorrect units in UBC source: 1.d-12vmrmmr`

File: bld/namelist_files/namelist_definition.xml [hunk 10]
```diff
@@ -6839,53 +6789,110 @@ if {{ hilight }}sulf_type{{ closehilight }} is 'FIXED'.
...
+<entry id="ubc_specifier" type="char*64(pcnst)"  category="upper_boundary"
+       group="upper_bc_opts" valid_values="" >
+
+List of fields that have fixed upper boundary conditions (temperature and mass mixing ratios)
+and corresponding data sources. Each entry in this list is of the form:
+  FLD->source
+where source may be of the following:
+  MSIS
+  TGCM
+  SNOE
```
[the remark this answers is not in the document]
Comment (fvitt) [reply bld/namelist_files/namelist_definition.xml]: I added brief descriptions for these legacy waccm upper boundary setters

File: src/chemistry/utils/tracer_data.F90 [hunk 11]
```diff
@@ -582,6 +593,8 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
 
     enddo flds_loop
 
+    call pio_seterrorhandling(File%curr_fileid, PIO_INTERNAL_ERROR)
+
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/utils/tracer_data.F90]: Corrected in many places in tracer_data.F90

File: src/chemistry/mozart/upper_bc_file.F90 [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+  logical, protected :: upper_bc_file_specified = .false.
+
+  ! private data members
+  character(len=cx) :: ubc_file_path = 'NONE'
+  character(len=32) :: ubc_file_input_type = 'NONE'
+  integer :: ubc_file_cycle_yr  = -huge(1)
+  integer :: ubc_file_fixed_ymd = -huge(1)
+  integer :: ubc_file_fixed_tod = -huge(1)
+
+  type(trfld), pointer :: fields(:)
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/mozart/upper_bc_file.F90]: done

File: src/chemistry/mozart/upper_bc_file.F90 [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+    call mpi_bcast(ubc_file_input_type, len(ubc_file_input_type), mpi_character, masterprocid, mpicom, ierr)
+    if (ierr /= 0) call endrun(prefix//'mpi_bcast error : ubc_file_input_type')
+    call mpi_bcast(ubc_file_fixed_ymd, 1, mpi_integer, masterprocid, mpicom, ierr)
+    if (ierr /= 0) call endrun(prefix//'mpi_bcast error : ubc_file_fixed_ymd')
+    call mpi_bcast(ubc_file_fixed_tod, 1, mpi_integer, masterprocid, mpicom, ierr)
+    if (ierr /= 0) call endrun(prefix//'mpi_bcast error : ubc_file_fixed_tod')
+    call mpi_bcast(ubc_file_cycle_yr,  1, mpi_integer, masterprocid, mpicom, ierr)
+    if (ierr /= 0) call endrun(prefix//'mpi_bcast error : ubc_file_cycle_yr')
+
+    upper_bc_file_specified = .not. ubc_file_path == 'NONE'
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/mozart/upper_bc_file.F90]: Changed as suggested

File: src/chemistry/mozart/upper_bc_file.F90 [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+    if (masterproc) then
+       write(iulog,*) prefix//'upper_bc_file_specified: ',upper_bc_file_specified
+       write(iulog,*) prefix//'ubc_file_path = '//trim(ubc_file_path)
+       write(iulog,*) prefix//'ubc_file_input_type = '//trim(ubc_file_input_type)
+       write(iulog,*) prefix//'ubc_file_cycle_yr = ',ubc_file_cycle_yr
+       write(iulog,*) prefix//'ubc_file_fixed_ymd = ',ubc_file_fixed_ymd
+       write(iulog,*) prefix//'ubc_file_fixed_tod = ',ubc_file_fixed_tod
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/mozart/upper_bc_file.F90]: Changes concats to commas

File: src/chemistry/mozart/upper_bc_file.F90 [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+    allocate(hist_names(num_ubc_flds), stat=ierr)
+    if (ierr /= 0) call endrun(prefix//'allocate error : hist_names')
+    hist_names = ' '
+
+    do m = 1,num_ubc_flds
+
+       call cnst_get_ind(trim(fields(m)%fldnam), ndx, abort=.true.)
+
+       select case ( to_lower(trim(fields(m)%units)) )
+       case ('k','kg/kg','mmr')
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/mozart/upper_bc_file.F90]: Changed as suggested

File: src/chemistry/mozart/upper_bc_file.F90 [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+          ubc_fact(m) = 1._r8
+       case ('mol/mol','mole/mole','vmr')
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/mozart/upper_bc_file.F90]: changed as suggested

File: src/chemistry/mozart/upper_bc.F90 [hunk 17]
```diff
@@ -157,39 +220,164 @@ subroutine ubc_init()
...
+             snoe_active = .true.
+          else
+             call endrun(prefix//'SNOE is not allowed in this configuration')
+          end if
+       else if (trim(ubc_source(m))=='ubc_file') then
+          file_spc_ndx(mm) = spc_ndx(m)
+          mm = mm+1
+       else
+          mmrndx = index(trim(ubc_source(m)),'mmr')
+          vmrndx = index(trim(ubc_source(m)),'vmr')
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/mozart/upper_bc.F90:314]: Yes

File: doc/ChangeLog (hunk 8 above)
Comment (fvitt) [reply doc/ChangeLog:17]: Added NAs

File: doc/ChangeLog (hunk 9 above)
Comment (fvitt) [reply doc/ChangeLog]: Removed the details here

**Review by peverwhee — APPROVED** [review]
looks good to me - one small typo

File: bld/namelist_files/namelist_definition.xml [hunk 18]
```diff
@@ -6839,53 +6789,112 @@ if {{ hilight }}sulf_type{{ closehilight }} is 'FIXED'.
...
+
+List of fields that have fixed upper boundary conditions (temperature and mass mixing ratios)
+and corresponding data sources. Each entry in this list is of the form:
+  FLD->source
+where source may be of the following:
+  MSIS - NRL neutral atmosphere empirical model (Mass Spectrometer Incoherent Scatter radar)
+  TGCM - Input file generated by Thermospheric General Circulation Model
+  SNOE - Input file generated from Student Nitric Oxide Explorer satellite observations
+  UBC_FILE - Generalized input file which spans the upper boundary altitude
+  value (mmr|vmr)
```
Comment (gold2718) [inline bld/namelist_files/namelist_definition.xml:6845]: This is better, thanks. Can you include a one-liner for `value` as well?
```suggestion
  value (mmr|vmr) - Set to a constant value (e.g., 0.123 mmr)
```
Is my example the right way to do this?

File: bld/namelist_files/namelist_definition.xml (hunk 10 above)
[the remark this answers is not in the document]
Comment (gold2718) [reply bld/namelist_files/namelist_definition.xml]: Okay but please see my other (new) comment about 'value'

**Review by gold2718 — APPROVED** [review]
Minor documentation request for namelist_definition (and please fix the type @peverwhee found) but otherwise, looks okay.

File: src/chemistry/mozart/upper_bc.F90 [hunk 19]
```diff
@@ -1,150 +1,212 @@
...
 !
-  public :: ubc_defaultopts    ! set default values of namelist variables
-  public :: ubc_setopts        ! get namelist input
+  public :: ubc_readnl         ! read namelist options for UBCs
   public :: ubc_init           ! global initialization
   public :: ubc_timestep_init  ! time step initialization
   public :: ubc_get_vals       ! get ubc values for this step
+  public :: ubc_get_flxs       ! get ub fluxes for this step
+  public :: ubc_fixed_conc     ! returns true for constituents that have fixed UBC
+  public :: ubc_fixed_temp     ! true if temperature at upper bounary is fixed
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/mozart/upper_bc.F90]: fixed

File: bld/namelist_files/namelist_definition.xml (hunk 18 above)
Comment (fvitt) [reply bld/namelist_files/namelist_definition.xml]: Made this change -- just did not like your example value.

**Pull request merged**